handler/src/main/java/io/netty/handler/ssl/ReleasableOpenSslEngine.java

```diff
@@ -0,0 +1,1988 @@
+            if (handshakeState != HandshakeState.FINISHED) {
+                if (handshakeState != HandshakeState.STARTED_EXPLICITLY) {
+                    // Update accepted so we know we triggered the handshake via wrap
+                    handshakeState = HandshakeState.STARTED_IMPLICITLY;
+                }
+
+                status = handshake();
+                if (status == NEED_WRAP) {
+                    return NEED_WRAP_OK;
+                }
+                if (engineClosed) {
+                    return NEED_WRAP_CLOSED;
+                }
+            }
+
+            // Write encrypted data to network BIO
+            int bytesConsumed = 0;
+            if (srcsOffset < srcsEndOffset) {
+                do {
+                    ByteBuffer src = srcs[srcsOffset];
+                    int remaining = src.remaining();
+                    if (remaining == 0) {
+                        // We must skip empty buffers as BIO_write will return 0 if asked to write something
+                        // with length 0.
+                        srcsOffset++;
+                        continue;
+                    }
+                    int written = writeEncryptedData(src);
+                    if (written > 0) {
+                        bytesConsumed += written;
+
+                        if (written == remaining) {
+                            srcsOffset++;
+                        } else {
+                            // We were not able to write everything into the BIO so break the write loop as otherwise
+                            // we will produce an error on the next write attempt, which will trigger a SSL.clearError()
+                            // later.
+                            break;
+                        }
+                    } else {
+                        // BIO_write returned a negative or zero number, this means we could not complete the write
+                        // operation and should retry later.
+                        // We ignore BIO_* errors here as we use in memory BIO anyway and will do another SSL_* call
+                        // later on in which we will produce an exception in case of an error
+                        SSL.clearError();
+                        break;
+                    }
+                } while (srcsOffset < srcsEndOffset);
+            }
+
+            // Number of produced bytes
+            int bytesProduced = 0;
+
+            if (capacity > 0) {
+                // Write decrypted data to dsts buffers
+                int idx = dstsOffset;
+                while (idx < endOffset) {
+                    ByteBuffer dst = dsts[idx];
+                    if (!dst.hasRemaining()) {
+                        idx++;
+                        continue;
+                    }
+
+                    int bytesRead = readPlaintextData(dst);
+
+                    // TODO: We may want to consider if we move this check and only do it in a less often called place
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)